src/web/auth.rs

```diff
@@ -0,0 +1,106 @@
+mod jwt;
+
+use std::{error, fmt};
+
+use rocket::http::Status;
+use rocket::outcome::Outcome;
+use rocket::request::{self, FromRequest, Request};
+use serde::Deserialize;
+
+pub use jwt::JwtValidator;
+
+#[derive(Debug, PartialEq, Deserialize)]
+pub enum Permission {
+    /// Allows superusers to perform any actions.
+    #[serde(rename(deserialize = "admin"))]
+    Admin,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Role {
+    /// Authorized user. Can create, retrieve, update, and delete private
+    /// snippets. May have additional permissions (e.g. if this is an admin
+    /// user).
+    Authorized {
+        name: String,
+        permissions: Vec<Permission>,
+    },
+
+    /// Anonymous user. Can create and retrieve publicly available snippets.
+    Guest,
+}
```

Comment:
I gave it a second thought and I think there's something odd with our approach. Roles normally are sets of permissions. They only make sense when we assign them to users. You know, just to make it easier to manage users by assigning them roles rather than sets of atomic permissions.

In our approach, however, a JWT token carries atomic permissions, rather than roles. That's why there's something I want to bring up for discussion.

1. The way how things currently are, I think it'd make sense to rename `Role` into `User` because it seems we will always have 2 states: anonymous and authenticated users.
2. We may end up having quite a lot of atomic permissions. Do you think it's ok to manage them on Auth0 side and send them over to API? Or should JWT tokens carry over roles, and maintain roles -> permissions mapping on API side (via some sort of configuration or something)?
3. nit: There's a slight difference between authentication and authorization.The latter implies that permissions to an action is granted. I wonder, would it make sense to use `Authenticated` instead of `Authorized`. WDYT?

Reply:
Thank you for the feedback! These are all good points.

> The way how things currently are, I think it'd make sense to rename Role into User because it seems we will always have 2 states: anonymous and authenticated users.

Agreed. I think my choice of names for this struct was very poor and created confusion (both `Role` and `Permission::Admin`).
 
> We may end up having quite a lot of atomic permissions. Do you think it's ok to manage them on Auth0 side and send them over to API? Or should JWT tokens carry over roles, and maintain roles -> permissions mapping on API side (via some sort of configuration or something)?

That's a good question. Auth0 has notions of permissions and roles. The latter is just a way to group a set of permissions. What is communicated to us via JWT tokens is permissions. I think we have two options:

* treat permissions in Auth0 as roles in XSnippet API and then map them to permissions that are only defined in the code
* continue using permissions configured in Auth0

To be honest, I don't think we are going to have that many individual permissions, even though I'm adding one more in #130. So I suggest we stick to the latter for the time being. The downside of this approach is that we need to define permissions in two places and keep them in sync.

> nit: There's a slight difference between authentication and authorization.The latter implies that permissions to an action is granted. I wonder, would it make sense to use Authenticated instead of Authorized. WDYT?

Sure! Sounds good!